src/libYARP_companion/src/yarp/companion/impl/Companion.cmdStats.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2006-2020 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * BSD-3-Clause license. See the accompanying LICENSE file for details.
+ */
+
+#include <yarp/companion/impl/Companion.h>
+
+#include <yarp/os/Bottle.h>
+#include <yarp/os/BufferedPort.h>
+#include <yarp/os/Contactable.h>
+#include <yarp/os/LogStream.h>
+#include <yarp/os/Network.h>
+#include <yarp/os/Property.h>
+#include <yarp/os/Semaphore.h>
+#include <yarp/os/Stamp.h>
+#include <yarp/os/TypedReaderCallback.h>
+#include <yarp/os/Value.h>
+
+using yarp::companion::impl::Companion;
+using yarp::os::Bottle;
+using yarp::os::BufferedPort;
+using yarp::os::Contactable;
+using yarp::os::LogStream;
+using yarp::os::NetworkBase;
+using yarp::os::Property;
+using yarp::os::Semaphore;
+using yarp::os::Stamp;
+using yarp::os::TypedReaderCallback;
+using yarp::os::Value;
+
+namespace {
+
+class CompanionStatsInput :
+        public TypedReaderCallback<Bottle>
+{
+public:
+    Contactable *port{nullptr};
+    std::mutex mutex;
+
+private:
+    Bottle value;
+    size_t bottles_received=0;
+    size_t bytes_received=0;
+    Stamp stamp;
+
+public:
+    CompanionStatsInput() = default;
+    void clearStats()
+    {
+        mutex.lock();
+        bottles_received = 0;
+        bytes_received = 0;
+        mutex.unlock();
+    }
+
+    void getStats(size_t& b1, size_t& b2)
+    {
+        mutex.lock();
+        b1= bytes_received;
+        b2= bottles_received;
+        mutex.unlock();
+    }
+
+    void init(Contactable& port)
+    {
+        this->port = &port;
+    }
+
+    using yarp::os::TypedReaderCallback<Bottle>::onRead;
+    void onRead(Bottle& datum) override
+    {
+        mutex.lock();
+        value = datum;
+        bottles_received++;
+        size_t  raw_bottle_bytes_c=0;
+        const char*   raw_bottle_bytes_p = value.toBinary(&raw_bottle_bytes_c);
+        bytes_received += raw_bottle_bytes_c;
+        port->getEnvelope(stamp);
+        mutex.unlock();
+    }
+};
+} // namespace
+
+int Companion::cmdStats(int argc, char *argv[])
+{
+    BufferedPort<Bottle >  localPort;
```

Comment:
Actually, I think this should not use `BufferedPort`, and use `Port` instead, `BufferedPort` will drop data, which means the statistics will wrong...

Reply:
On the contrary, `yarp::os::Port` may not lose packets, but it slows down the communication, altering the statistics as well.
Since the solution is really dependent on the specific use case, we decided to allow the user to choose the preferred way to receive data, using a parameter to select `BufferedPort`  or `Port`